// @aws-cdk/assets/lib/staging.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Staging = void 0;
const jsiiDeprecationWarnings = require("../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const core_1 = require("@aws-cdk/core");
const compat_1 = require("./compat");
/**
 * Deprecated
 * @deprecated use `core.AssetStaging`
 */
class Staging extends core_1.AssetStaging {
    constructor(scope, id, props) {
        super(scope, id, {
            sourcePath: props.sourcePath,
            exclude: props.exclude,
            ignoreMode: props.ignoreMode,
            extraHash: props.extraHash,
            follow: compat_1.toSymlinkFollow(props.follow),
        });
        try {
            jsiiDeprecationWarnings.print("@aws-cdk/assets.Staging", "use `core.AssetStaging`");
            jsiiDeprecationWarnings._aws_cdk_assets_StagingProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.constructor);
            }
            throw error;
        }
    }
}
exports.Staging = Staging;
_a = JSII_RTTI_SYMBOL_1;
Staging[_a] = { fqn: "@aws-cdk/assets.Staging", version: "1.153.1" };
